neo/Network/neonetwork/network/nodemanager.py

```diff
@@ -0,0 +1,382 @@
+import asyncio
+import socket
+import traceback
+import errno
+from contextlib import suppress
+from datetime import datetime
+from functools import partial
+from socket import AF_INET as IP4_FAMILY
+from typing import Optional, List
+
+from neo.Core.TX.Transaction import Transaction as OrigTransaction
+from neo.Network.neonetwork.common import msgrouter, wait_for
+from neo.Network.neonetwork.common.singleton import Singleton
+from neo.Network.neonetwork.network import utils as networkutils
+from neo.Network.neonetwork.network.mempool import MemPool
+from neo.Network.neonetwork.network.node import NeoNode
+from neo.Network.neonetwork.network.protocol import NeoProtocol
+from neo.Network.neonetwork.network.relaycache import RelayCache
+from neo.Network.neonetwork.network.requestinfo import RequestInfo
+from neo.Settings import settings
+from neo.logging import log_manager
+
+logger = log_manager.getLogger('network')
+
+
+# log_manager.config_stdio([('network', 10)])
+
+
+class NodeManager(Singleton):
+    PEER_QUERY_INTERVAL = 15
+    NODE_POOL_CHECK_INTERVAL = 2.5 * PEER_QUERY_INTERVAL  # this allows for enough time to get new addresses
+
+    ONE_MINUTE = 60
+
+    MAX_ERROR_COUNT = 5  # maximum number of times adding a block or header may fail before we disconnect it
+    MAX_TIMEOUT_COUNT = 15  # maximum count the node responds slower than our threshold
+
+    MAX_NODE_POOL_ERROR = 2
+    MAX_NODE_POOL_ERROR_COUNT = 0
+
+    # we override init instead of __init__ due to the Singleton (read class documentation)
+    def init(self):
+        self.loop = asyncio.get_event_loop()
+        self.max_clients = settings.CONNECTED_PEER_MAX
+        self.min_clients = settings.CONNECTED_PEER_MIN
+        self.id = id(self)
+        self.mempool = MemPool()
+
+        # a list of nodes that we're actively using to request data from
+        self.nodes = []  # type: List[NeoNode]
+        # a list of host:port addresses that have a task pending to to connect to, but are not fully processed
+        self.queued_addresses = []
+        # a list of addresses which we know are bad. Reasons include; failed to connect, went offline, poor performance
+        self.bad_addresses = []
+        # a list of addresses that we've tested to be alive but that we're currently not connected to because we've
+        # reached our `max_clients` setting. We use these addresses to quickly replace a bad node
+        self.known_addresses = []
+
+        self.connection_queue = asyncio.Queue()
+
+        # a list for gathering tasks such that we can manually determine the order of shutdown
+        self.tasks = []
+        self.shutting_down = False
+
+        self.relay_cache = RelayCache()
+
+        msgrouter.on_addr += self.on_addr_received
+
+        self.running = False
+
+    async def start(self):
+        host = 'localhost'
+        port = 8888  # settings.NODE_PORT
```

Comment:
why is `settings.NODE_PORT` replaced here?

Reply:
Just a local test thing that slipped in. The default port conflicts with my privnet

Reply:
- [ ] reminder to revert to `settings.NODE_PORT` prior to merging

Reply:
fixed in #948